internals/overlord/logstate/gatherer.go

```diff
@@ -0,0 +1,255 @@
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package logstate
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/canonical/pebble/internals/logger"
+	"github.com/canonical/pebble/internals/plan"
+	"github.com/canonical/pebble/internals/servicelog"
+	"gopkg.in/tomb.v2"
+)
+
+const parserSize = 4 * 1024
+
+var (
+	tickPeriod      = 1 * time.Second
+	teardownTimeout = 5 * time.Second
+)
+
+// logGatherer is responsible for collecting service logs from a bunch of
+// services, and sending them to its logClient.
+// One logGatherer will run per log target. Its loop() method should be run
+// in its own goroutine. It can be stopped in a separate goroutine by calling
+// the stop() method.
+// The logGatherer will "flush" the client:
+//   - on a regular cadence (e.g. every 1 second)
+//   - when it is told to shut down.
+//
+// Its client may also flush itself when its internal buffer reaches a certain
+// size.
+type logGatherer struct {
+	targetName string
+	tickPeriod time.Duration
+	client     logClient
+
+	entryCh     chan servicelog.Entry
+	pullers     map[string]*logPuller
+	pullersLock sync.Mutex
+
+	// Context to pass to client
+	// When we are asked to shutdown, we can cancel this context to ensure the
+	// client is not blocking us.
+	clientCtx    context.Context
+	clientCancel context.CancelFunc
+
+	// tomb for the main loop
+	tomb *tomb.Tomb
+}
+
+func newLogGatherer(target *plan.LogTarget) (*logGatherer, error) {
+	client, err := newLogClient(target)
+	if err != nil {
+		return nil, fmt.Errorf("could not create log client: %w", err)
+	}
+
+	g := &logGatherer{
+		targetName: target.Name,
+		tickPeriod: tickPeriod,
+		client:     client,
+		entryCh:    make(chan servicelog.Entry),
+		pullers:    map[string]*logPuller{},
+		tomb:       &tomb.Tomb{},
+	}
+	g.clientCtx, g.clientCancel = context.WithCancel(context.Background())
+	g.tomb.Go(g.loop)
+	return g, nil
+}
+
+func (g *logGatherer) planChanged(pl *plan.Plan, buffers map[string]*servicelog.RingBuffer) {
+	g.pullersLock.Lock()
+	defer g.pullersLock.Unlock()
+
+	newPullers := map[string]*logPuller{}
+
+	for _, service := range pl.Services {
+		target := pl.LogTargets[g.targetName]
+		if !service.LogsTo(target) {
+			continue
+		}
+
+		if puller, ok := g.pullers[service.Name]; ok {
+			// Move over existing puller
+			newPullers[service.Name] = puller
+			delete(g.pullers, service.Name)
+			continue
+		}
+
+		buffer, bufferExists := buffers[service.Name]
+		if !bufferExists {
+			// We don't yet have a reference to the service's ring buffer
+			// Need to wait until serviceStarted
+			continue
+		}
+
+		// Create new puller
+		newPullers[service.Name] = newLogPuller(g.entryCh, buffer)
+	}
+
+	// Old pullers for now-removed services need to be shut down.
+	for _, puller := range g.pullers {
+		_ = puller.tomb.Killf("plan changed")
```

Comment:
Should we debug-log this error?

Reply:
Killf semantics are confusing. The error returned is exactly the same obtained from the arguments you pass in - it doesn't indicate a problem running the method. I could change this to the following to avoid confusion?
```
puller.tomb.Kill(errors.New("plan changed"))
```